test: Assumeutxo: snapshots with less work should not be loaded

hernanmarino opened the pull request:
This PR adds a test which checks that snapshots with less accumulated work than the node's active chain, should not be loaded and return with an error. Although in a different context of discussion the missing test was detect in a thread in https://github.com/bitcoin/bitcoin/pull/29394 (see https://github.com/bitcoin/bitcoin/pull/29394#discussion_r1484122214)

hernanmarino committed:
test: Assumeutxo: snapshots with less work should not be loaded

aureleoules commented:
> Looks like [corecheck.dev/bitcoin/bitcoin/pulls/29428](https://corecheck.dev/bitcoin/bitcoin/pulls/29428) is dead?

@maflcko thanks for the ping. The code I added to filter out flaky coverage was too slow and isn't very reliable so I removed it. Unfortunately, the UI will now display flaky coverage.

hernanmarino commented:
> You'll have to remove the TODO on the top of the file as well, when the test was added?

Done. It'll be updated with my next push

BrandonOdiwuor reviewed (commented):
Code Review ACK 86037415dd83da34d829f2f6bfb8dcbe08968180

Should remove the TODO fixed by this commit

hernanmarino commented:
> Are you still working on this?

I haven´t been able to think of / work on other tests to add, but the current commit test is ready for review, so i am taking the PR out of Draft status.

hernanmarino commented:
Updated to remove a (no lonnger valid) TODO comment as requested in https://github.com/bitcoin/bitcoin/pull/29428#issuecomment-1952480346. Also updated the PR title and description.

hernanmarino commented:
> Please delete the hidden comment in the pull description, because this will end up in the merge commit.

Oh, sorry that. Deleted.

alfonsoromanz reviewed (approved):
Code review and tested ACK 4922f56660ae7834598c6dc4904a6711c23edfa3

alfonsoromanz commented on test/functional/feature_assumeutxo.py:
```diff
@@ -20,8 +20,6 @@
 - TODO: Valid snapshot file, but referencing a snapshot block that turns out to be
       invalid, or has an invalid parent
-- TODO: Valid snapshot file and snapshot block, but the block is not on the
-      most-work chain
```
Maybe the line that should be removed here is this one? `- TODO: A descendant of the snapshot block`. I believe the scenario where the snapshot block is not on the most-work chain is addressed here: https://github.com/bitcoin/bitcoin/pull/29996

hernanmarino replied:
> Why is this removed? The block is in the most-work chain. I think this comment meant a block that is _not_ in the most-work chain, no?

Maybe i missinterpreted the TODO comment meaning . To which TODO were you referring here  
https://github.com/bitcoin/bitcoin/pull/29428#issuecomment-1952480346 ?

hernanmarino replied:
Restored that TODO in my last push. Thanks

alfonsoromanz reviewed (approved):
Re ACK 2f1b1eee8b8a8f546fb5975ac529c4032e46f069

hernanmarino commented:
Updated to re add an incorrectly removed TODO comment.

rkrux reviewed (approved):
crACK, utACK [2f1b1eee](https://github.com/bitcoin/bitcoin/pull/29428/commits/2f1b1eee8b8a8f546fb5975ac529c4032e46f069)

Thanks for adding the test, it's short and to the point. However, I've not been able to build the branch because it fails on my system (MAC i7) with the following error. I understand it is because this branch doesn't contain the latest changes of the master branch, notably this one: https://github.com/bitcoin/bitcoin/pull/29081.

```
util/time.cpp:54:9: error: use of undeclared identifier 'gmtime_s'; did you mean 'gmtime_r'?
    if (gmtime_s(&epoch, &time_val) != 0) {
        ^~~~~~~~
        gmtime_r
```

achow101 commented:
Needs rebase (for some reason Drahtbot isn't detecting this)

hernanmarino commented:
> Needs rebase (for some reason Drahtbot isn't detecting this)

Rebased. Also deleted a couple of redundant comment as suggested by @maflcko to retrigger the false negative CI

furszy commented on test/functional/feature_assumeutxo.py:
```diff
@@ -161,6 +161,14 @@ def test_invalid_file_path(self):
+        assert_equal(node.getblockcount(), FINAL_HEIGHT)
+        with node.assert_debug_log(expected_msgs=["[snapshot] activation failed - work does not exceed active chainstate"]):
+            assert_raises_rpc_error(-32603, "Unable to load UTXO snapshot", node.loadtxoutset, dump_output_path)
+        self.restart_node(0, extra_args=self.extra_args[0])
```
why restart if you aren't doing anything with the node?

kevkevinpal commented:
utACK [df6dc2a](https://github.com/bitcoin/bitcoin/pull/29428/commits/df6dc2aaaeffc664006b86ee8c8797dc484ec40e)

alfonsoromanz reviewed (approved):
Re ACK df6dc2aaaeffc664006b86ee8c8797dc484ec40e. Make is successful and the test passes. 

<img width="500" alt="Screenshot 2024-05-28 at 09 55 15" src="https://github.com/bitcoin/bitcoin/assets/19962151/e3e974e8-a507-49c9-8d77-6457f6f95443">
<img width="500" alt="Screenshot 2024-05-28 at 09 58 39" src="https://github.com/bitcoin/bitcoin/assets/19962151/f2e020ed-efc6-4854-97b5-f919ac1e0cdf">

achow101 commented:
ACK df6dc2aaaeffc664006b86ee8c8797dc484ec40e
